server/resource_server/emails/verification/zk/circuit/mimc_circuit.go

```diff
@@ -0,0 +1,63 @@
+package circuit
+
+import (
+	"github.com/consensys/gnark/frontend"
+	"github.com/consensys/gnark/std/hash/mimc"
+	"github.com/consensys/gnark/std/math/bits"
+	"globe-and-citizen/layer8/server/resource_server/utils"
+)
+
+type MimcCircuit struct {
+	SaltAsVariables  [utils.EmailFrRepresentationSize]frontend.Variable `gnark:",public"`
+	EmailAsVariables [utils.EmailFrRepresentationSize]frontend.Variable `gnark:",secret"`
+
+	VerificationCode [utils.VerificationCodeSize]frontend.Variable `gnark:",public"`
+}
+
+func NewMimcCircuit() *MimcCircuit {
+	return new(MimcCircuit)
+}
+
+func (c *MimcCircuit) Define(api frontend.API) error {
+	mimcInstance, _ := mimc.NewMiMC(api)
+
+	for i := 0; i < utils.EmailFrRepresentationSize; i++ {
+		emailFrVariableBits := bits.ToBinary(api, c.EmailAsVariables[i])
+		saltFrVariableBits := bits.ToBinary(api, c.SaltAsVariables[i])
+
+		// representation for frElementsEmail[i] ^ frElementsSalt[i]
+		currentVariable := frontend.Variable(0)
+		powerOfTwo := frontend.Variable(1)
+		for j := 0; j < len(emailFrVariableBits); j++ {
```

Comment:
On reflection, I now fully appreciate why you are xor'ing. Without doing this, adding  the salt, anyone could spoof our implementation especially given that it is open source. Without knowing their salt, they cannot.

Reply:
Eventually though, we will need to improve on the source of the salt because to do SCRAM (Salted Challenge Response Authentication Mechanism), we actually SEND them their salt... obviously a problem in the long run. For now, a non-issue.

Reply:
Yes, we'll have to think about that. Probably it's worth generating another sequence of random bytes and to save it into the database when running email verification.